src/python/bot/fuzzers/blackbox/engine.py

```diff
@@ -0,0 +1,171 @@
+    --app_args: Arguments to pass to the test binary.
+    --input_dir: An optional corpus containing test data.
+    --output_dir: If specified, write generated test cases to this directory
+                  along with their output. This is specified during fuzzing but
+                  not reproduction. Only the crashes must be written, but
+                  crashes will be identified automatically if the fuzzer does
+                  not detect them itself. Tests should have the prefix "fuzz-",
+                  and their output should have the same name as the test but
+                  with the prefix "output-" instead of "fuzz-".
+    --testcase_path: If specified, a path to a test case to reproduce. Output
+                     should be written directly to stdout or stderr.
+  """
+
+  @property
+  def name(self):
+    return 'blackbox'
+
+  def prepare(self, corpus_dir, target_path, build_dir):  # pylint: disable=unused-argument
+    """Prepare for a fuzzing session by generating options.
+
+    Though blackbox fuzzers follow the engine interface, they must be launched
+    in a different manner from most other engine fuzzers. Instead of running a
+    target directly, these fuzzers tend to be wrapper scripts which generate
+    test cases and pass them to a binary that is managed by the infrastructure.
+
+    Args:
+      corpus_dir: The main corpus directory.
+      target_path: Path to the target.
+      build_dir: Path to the build directory.
+    Returns:
+      A FuzzOptions object.
+    """
+    app_path = environment.get_value('APP_PATH')
+    app_args = testcase_manager.get_command_line_for_application(
+        get_arguments_only=True).strip()
+    arguments = [
+        f'--app_path={app_path}',
+        f'--app_args={app_args}',
+        f'--input_dir={corpus_dir}',
+    ]
+    return engine.FuzzOptions(corpus_dir, arguments, {})
+
+  def fuzz(self, target_path, options, reproducers_dir, max_time):
+    """Run a fuzzing session.
+    Args:
+      target_path: Path to the target.
+      options: The FuzzOptions object returned by prepare().
+      reproducers_dir: The directory to put reproducers in when crashes
+          are found.
+      max_time: Maximum allowed time for the fuzzing to run.
+   Returns:
+      A FuzzResult object.
+    """
+    os.chmod(target_path, 0o775)
+
+    args = options.arguments
+    args.append(f'--output_dir={reproducers_dir}')
+
+    runner = new_process.UnicodeProcessRunner(target_path)
+    result = runner.run_and_wait(timeout=max_time, additional_args=args)
+
+    crashes = []
+    for testcase_path in os.listdir(reproducers_dir):
+      if not testcase_path.startswith(TESTCASE_PREFIX):
+        continue
+
+      output_path = OUTPUT_PREFIX + testcase_path[len(TESTCASE_PREFIX):]
+      absolute_output_path = os.path.join(reproducers_dir, output_path)
+
+      # If no output was written for a test case, skip it.
+      if not os.path.exists(absolute_output_path):
+        continue
```

Comment:
+1, how is this backwards compatible, or is it not possible to do ?

Reply:
At a high level it would just be a matter of moving some of the logic we have for test case processing and running to the fuzz function. That would also let us share the post-processing logic in a cleaner way, but I don't think it should be a blocker for the initial version.

If we were to change the existing blackbox pipeline to write the output to files in the same way we're doing here and port it over (have some environment variable set which determines whether we run in this new way or have CF process everything) it should be fairly easy to share the logic between both. I'd like to start with this for the case where the fuzzer processes test cases on its own for now, but I don't see any inherent problems with extending this to work with the old pipeline as well.

Reply:
I'm more envisioning just having every single blackbox fuzzer go through this pipeline, and completely deleting the old one. We shouldn't need to maintain 3 separate fuzzing pipelines.

Is there any problem with CF just always verifying/processing testcase in the output directory? The blackbox fuzzer can always do its own processing as well to trim down on results etc. That way it will be perfectly backwards compatible. As is, this PR expects the fuzzer to always process testcases since it expects this output file. We should relax this requirement.

Reply:
+1 to what @oliverchang said.

Reply:
I'm having trouble finding a clean way to support this in this change without either duplicating or, more likely, re-implementing a lot of the existing test case processing logic, but I agree that we should do this. Here's my current plan, since this CL has been quiet for a bit too long:

1. I'm going to include the usage in fuzz task in this change to pull all of this together. With that, this should actually be usable after this PR lands. I'll include a few TODOs along with that which I think will make the next steps for transitioning a bit clearer.
2. If an environment variable is set, we'll use this new pipeline for blackbox fuzzers which are capable of processing test cases on their own.
3. Once this seems stable, I'll convert the entire blackbox fuzzing pipeline to use the engine interface. This will avoid some of the code duplication issues we've been running into, and resolve the backwards compatibility issues noted here.
4. We should be able to clean up some of the other code in follow up changes, such as the use of different crash types between the two interface types, and hopefully get rid of some of the branching logic for handling both cases (though I'm not totally sure the latter will be easy to resolve).

WDYT?